src/plugins/intel_npu/src/plugin/npuw/llm_prefix_caching.hpp

```diff
@@ -0,0 +1,231 @@
+
+#pragma once
+
+#include <list>
+#include <memory>
+#include <mutex>
+#include <regex>
+#include <unordered_map>
+
+#include "openvino/runtime/icompiled_model.hpp"
+#include "openvino/runtime/itensor.hpp"
+#include "openvino/runtime/so_ptr.hpp"
+
+namespace ov {
+namespace npuw {
+
+/**
+ * @brief KV cache data container for storing tensor pairs per layer
+ *
+ * This type alias defines a container that stores KV cache tensors organized by layer.
+ * Each element in the vector represents one layer's KV cache data as a pair of:
+ * - string: Tensor name/identifier (e.g., "past_key_values.0.key", "past_key_values.0.values")
+ * - ov::SoPtr<ov::ITensor>: Smart pointer to the actual KV cache tensor data
+ *
+ * The container holds all KV cache tensors for a specific block of tokens,
+ * enabling efficient storage and retrieval of cached computation results
+ * in the prefix caching system.
+ */
+using KVData = std::vector<std::pair<std::string, ov::SoPtr<ov::ITensor>>>;
+
+class KVBlock {
+public:
+    KVBlock(size_t block_size)
+        : m_block_size(block_size),
+          m_ref_count(0),
+          m_is_full(false),
+          m_token_start(0),
+          m_block_hash(0),
+          m_parent_block_hash(0) {
+        m_token_hashes.reserve(m_block_size);
+    }
+
+    bool is_full() const {
+        return m_is_full;
+    }
+
+    void set_token_start(size_t token_start) {
+        m_token_start = token_start;
+    }
+
+    size_t get_token_start() const {
+        return m_token_start;
+    }
+
+    uint64_t get_block_hash() const {
+        return m_block_hash;
+    }
+
+    const std::unordered_set<uint64_t>& get_child_block_hashes() const {
+        return m_child_block_hashes;
+    }
+
+    uint64_t get_parent_block_hash() const {
+        return m_parent_block_hash;
+    }
+
+    const KVData& get_block_kv_data() const {
+        return m_kv_data;
+    }
+
+    /**
+     * @brief Add the entire block's token data
+     * @param token_hashes Hash values of all tokens in the block
+     * @param kv_tensors KV cache data for all tokens in the block
+     * @return Whether the addition was successful
+     */
+    bool add_block(const std::vector<uint64_t>& token_hashes, const KVData& kv_tensors);
+    void link_blocks(std::shared_ptr<KVBlock> prev_block);
+    void unlink_blocks(std::shared_ptr<KVBlock> prev_block);
+    void print_block_info(bool verbose) const;
+
+private:
+    /**
+     * @brief Compute the block's hash value
+     * @param token_hashes Hash values of all tokens in the block
+     * @return The block's hash value
+     *
+     * Note: The block hash is derived from the last token hash in the sequence,
+     * as each token hash is calculated based on preceding tokens, ensuring
+     * a cumulative representation of the block's content.
+     */
+    uint64_t compute_block_hash(const std::vector<uint64_t>& token_hashes) const;
+
+    size_t m_block_size;
+    std::vector<uint64_t> m_token_hashes;
+    size_t m_ref_count;
```

Comment:
what this ref_count stands for if (looking at the code below) the PrefixCacheManager operates with `shared_ptr<KVBlock>`?